Sources/SwiftFusion/Core/Tuple+EuclideanVectorN.swift

```diff
@@ -0,0 +1,159 @@
+// Copyright 2020 The SwiftFusion Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import PenguinStructures
+
+extension Empty: AdditiveArithmetic {
+  public static func + (_ lhs: Self, _ rhs: Self) -> Self {
+    return Empty()
+  }
+  public static func - (_ lhs: Self, _ rhs: Self) -> Self {
+    return Empty()
+  }
+  public static var zero: Self {
+    return Empty()
+  }
+}
+
+extension Tuple: AdditiveArithmetic where Head: AdditiveArithmetic, Tail: AdditiveArithmetic {
+  public static func + (_ lhs: Self, _ rhs: Self) -> Self {
+    return Tuple(head: lhs.head + rhs.head, tail: lhs.tail + rhs.tail)
+  }
+  public static func - (_ lhs: Self, _ rhs: Self) -> Self {
+    return Tuple(head: lhs.head - rhs.head, tail: lhs.tail - rhs.tail)
+  }
+  public static var zero: Self {
+    return Tuple(head: Head.zero, tail: Tail.zero)
+  }
+}
+
+extension Empty: Differentiable {
+  public typealias TangentVector = Self
+  public mutating func move(along direction: TangentVector) {}
+}
+
+extension Tuple: Differentiable
+where Head: Differentiable, Tail: Differentiable, Tail.TangentVector: TupleProtocol {
+  public typealias TangentVector = Tuple<Head.TangentVector, Tail.TangentVector>
+  public mutating func move(along direction: TangentVector) {
+    head.move(along: direction.head)
+    tail.move(along: direction.tail)
+  }
+}
+
+extension Empty: EuclideanVector {
+  @differentiable
+  public static func += (_ lhs: inout Self, _ rhs: Self) {}
+
+  @differentiable
+  public static func -= (_ lhs: inout Self, _ rhs: Self) {}
+
+  // MARK: - Scalar multiplication.
+
+  @differentiable
+  public static func *= (_ lhs: inout Self, _ rhs: Double) {}
+
+  // MARK: - Euclidean structure.
+
+  /// The inner product of `self` with `other`.
+  @differentiable
+  public func dot(_ other: Self) -> Double { return 0 }
+
+  // MARK: - Conversion to/from a concrete dynamically-sized `Vector`.
+
+  /// Creates an instance with the same elements as `vector`.
+  @differentiable
+  public init(_ vector: Vector) { self.init() }
+
+  /// A `Vector` containing the same elements as `self`.
+  @differentiable
+  public var vector: Vector { return Vector([]) }
+}
+
+extension Empty: EuclideanVectorN {
```

Comment:
This `EuclideanVectorN` is the "fixed size vector" protocol that exists in SwiftFusion now. It has some problems, but I think it's good enough to get the generic factor graph stuff working, so I'm going to use it in all the generic factor graph stuff and we can replace it later when we have more time to figure out fixed size vectors.